test/test_pebble.py

```diff
@@ -208,13 +211,13 @@ def test_warning_init(self):
         self.assertEqual(warning.repeat_after, '2s')
 
     def test_warning_from_dict(self):
-        d = {
+        d: pebble._WarningDict = {
```

Comment:
Hmm, interesting. This is probably my bad, but because those `from_dict` methods are part of the public API, it seems to me like their types should be also. So perhaps we should remove the underscores from all of the following? (And move them to the top of pebble.py with the other public typed dicts.)

```
ChangeDict
CheckInfoDict
FileInfoDict
ProgressDict
ServiceInfoDict
SystemInfoDict
TaskDict
WarningDict
```

Thoughts?

Reply:
Or maybe we should leave that till we have access to `Required` and `NotRequired`, without which these are kind of awkward?

Reply:
> because those `from_dict` methods are part of the public API, it seems to me like their types should be also.

I find this a compelling argument. It does seem like if you want to use the `from_dict` methods, and also have strict type checking, then you're forced to use the various `_*Dict` types.

> So perhaps we should remove the underscores from all of the following? (And move them to the top of pebble.py with the other public typed dicts.)

I guess my only question here is when you expect people to use the `from_dict` methods (for `Change`, `CheckInfo`, `FileInfo`, `Progress`, `ServiceInfo`, `SystemInfo`, `Task`, `Warning`). These are all representing information that flows unidirectionally from Pebble, right? Would a charmer (or more broadly someone who is using ops.pebble to manage pebble outside of charming) ever have reason to construct one of those objects (specifically from a dict, but really in general)?

Or more succinctly, is the issue really that `from_dict()` should be `_from_dict()` for that list of objects? And the type being private would then match? (Maybe this horse has bolted anyway, but I would be interested to know the answer).

Reply:
> Or maybe we should leave that till we have access to `Required` and `NotRequired`, without which these are kind of awkward?

I missed that we already had `typing_extensions` in our requirements-dev, and already use `Required` elsewhere. So I can clean those up.

Reply:
Nice!

Reply:
> Or more succinctly, is the issue really that from_dict() should be _from_dict() for that list of objects? And the type being private would then match? (Maybe this horse has bolted anyway, but I would be interested to know the answer).

Yeah, I think you're right -- these should have been `_from_dict` in the first place, as that's an internal implementation detail, and we don't really expect users to instantiate these because they're from responses from the Pebble server.

But yeah, this cat is already out of the bag (how many other idioms are there for that? :-). So I think we should go all in and remove the underscore. We don't have to add them to anyone's `__all__`, as `ops.pebble` stuff isn't imported to the top level (intentionally), so it won't pollute namespaces.

Reply:
As discussed, I'll open a separate PR for this.